fix(helm): helm pull --repo to use repo creds from helm repos file

andreaskaris opened the pull request:
Up to this point, helm pull --repo required --username and --password
even if the repository had already been added via helm repo add.
    
From now on, check repositories in the repositories file and if the URL
matches, pull the chart with the username and password from the entry.

Fixes #9599

Signed-off-by: Andreas Karis <ak.karis@gmail.com>

**What this PR does / why we need it**:

Set up the following reproducer - set up the following directory structure for httpd:
```
yum install httpd -y
mkdir cd /var/www/html/charts
cd /var/www/html/charts
helm repo add brigade https://brigadecore.github.io/charts
helm pull brigade/brigade
helm pull brigade/kashti
helm repo index . --url http://127.0.0.1/charts
htpasswd -c .htpasswd testuser   # --> "password"
cat <<'EOF' > .htaccess
AuthType Basic
AuthName "Restricted Access"
AuthUserFile /var/www/html/charts/.htpasswd
Require valid-user
EOF
restorecon -R .

vi /etc/httpd/conf/httpd.conf
# <Directory "/var/www/html/charts"> 
#    AllowOverride AuthConfig
# </Directory>
```

Now, the repo should be password protected:
```
curl -u testuser:password  http://127.0.0.1/charts/index.yaml
```

```
# helm repo add localhost http://127.0.0.1/charts/ --username testuser --password password
# helm search repo localhost
NAME             	CHART VERSION	APP VERSION	DESCRIPTION                                       
localhost/brigade	1.10.0       	v1.5.0     	Brigade provides event-driven scripting of Kube...
localhost/kashti 	0.7.0        	v0.4.0     	A Helm chart for Kubernetes
# helm pull localhost/brigade
# ll | grep brigade
-rw-r--r--. 1 root root 14278 Feb 10 15:01 brigade-1.10.0.tgz
```

However, with --repo, this won't work:
```
# helm pull --repo http://127.0.0.1/charts/ brigade
Error: failed to fetch http://127.0.0.1/charts/brigade-1.10.0.tgz : 401 Unauthorized
```

The username and password must be provided:
```
# helm pull --repo http://127.0.0.1/charts/ brigade
Error: looks like "http://127.0.0.1/charts/" is not a valid chart repository or cannot be reached: failed to fetch http://127.0.0.1/charts/index.yaml : 401 Unauthorized
```

After applying the fix:
```
[root@fedora1 helm]# /home/akaris/development/helm/bin/helm pull --repo http://127.0.0.1/charts/ brigade; if [ -f brigade-1.10.0.tgz ]; then echo ok; rm -f brigade-1.10.0.tgz;fi
ok
[root@fedora1 helm]# /home/akaris/development/helm/bin/helm pull --repo http://127.0.0.1/charts brigade; if [ -f brigade-1.10.0.tgz ]; then echo ok; rm -f brigade-1.10.0.tgz;fi
ok
[root@fedora1 helm]# /home/akaris/development/helm/bin/helm pull --repo http://127.0.0.1/charts//// brigade; if [ -f brigade-1.10.0.tgz ]; then echo ok; rm -f brigade-1.10.0.tgz;fi
ok
[root@fedora1 helm]# /home/akaris/development/helm/bin/helm pull --repo http://127.0.0.1///charts//// brigade; if [ -f brigade-1.10.0.tgz ]; then echo ok; rm -f brigade-1.10.0.tgz;fi
ok
[root@fedora1 helm]# /home/akaris/development/helm/bin/helm pull --repo http://127.0.0.1/charts/a//// brigade; if [ -f brigade-1.10.0.tgz ]; then echo ok; rm -f brigade-1.10.0.tgz;fi
Error: looks like "http://127.0.0.1/charts/a////" is not a valid chart repository or cannot be reached: failed to fetch http://127.0.0.1/charts/a////index.yaml : 401 Unauthorized
```



<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**Special notes for your reviewer**:

**If applicable**:
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

andreaskaris committed:
fix(helm): helm pull --repo to use repo creds from helm repos file

Up to this point, helm pull --repo required --username and --password
even if the repository had already been added via helm repo add.

From now on, check repositories in the repositories file and if the URL
matches, pull the chart with the username and password from the entry.

Fixes #9599

Co-authored-by: Andreas Karis <ak.karis@gmail.com>
Co-authored-by: Felipe Santos <felipecassiors@gmail.com>
Signed-off-by: Andreas Karis <ak.karis@gmail.com>

felipecrs commented:
I can't assess much from the technical perspective here, my only concern is:

When we run `helm dep update`, the exact same thing occurs (Helm tries to find repos with credentials already set in which matches the repo url set in the dependencies). So maybe, this does not need to be reimplemented, but reused from the `helm dep update`.

andreaskaris commented:
The code for `helm dep update` that's in charge of this is here:
https://github.com/helm/helm/blob/e87f81563788f311e7a5a7640385c003d5d6096b/pkg/downloader/manager.go#L243

and that calls findChartURL:
https://github.com/helm/helm/blob/e87f81563788f311e7a5a7640385c003d5d6096b/pkg/downloader/manager.go#L689

It then uses `DownloadTo` to download the chart:
https://github.com/helm/helm/blob/e87f81563788f311e7a5a7640385c003d5d6096b/pkg/downloader/manager.go#L355

These methods are all private to the `downloader` package so they can't be reused.

The suggested code merely browses the repository .yaml file, looks for the first match for the repository's URL (exact match) and then returns the repository name. It then prepends the repository name to the chart name and then calls `c.DownloadTo` as if the user had called this with `repository_name`/`chart_name`. So it mostly recycles existing logic.

The alternative would be to fetch the username and password from the file in `p.Settings.RepositoryConfig` and to set them in `repo.FindChartInAuthAndTLSRepoURL`   instead of `p.Username` and `p.Password`.

felipecrs commented:
Methods being private does not mean they their access can not be changed, right? But anyway, your approach does the trick for me, as I said, I can't assess much from the implementation perspective.

andreaskaris commented:
I'm not a maintainer here, either. I'm just looking into helm and I'm chasing for some low-hanging fruit bug reports.

I changed it a bit and I think its cleaner now.

This now checks if username and password are both not provided and if they are not provided, then it will attempt to find a repository entry in the RepositoryConfig file and it will use the username and password which are provided there.

Before:
~~~
[akaris@linux helm]$ helm pull  --repo http://localhost:8080 test
Error: looks like "http://localhost:8080" is not a valid chart repository or cannot be reached: failed to fetch http://localhost:8080/index.yaml : 401 Unauthorized
[akaris@linux helm]$ 
[akaris@linux helm]$ helm pull  --repo http://localhost:8080 test --username test
Error: looks like "http://localhost:8080" is not a valid chart repository or cannot be reached: failed to fetch http://localhost:8080/index.yaml : 401 Unauthorized
[akaris@linux helm]$ 
[akaris@linux helm]$ helm pull  --repo http://localhost:8080 test --username test --password test2
Error: looks like "http://localhost:8080" is not a valid chart repository or cannot be reached: failed to fetch http://localhost:8080/index.yaml : 401 Unauthorized
[akaris@linux helm]$ 
[akaris@linux helm]$ helm pull  --repo http://localhost:8080 test --username test --password test
[akaris@linux helm]$ rm -f test-0.1.0.tgz 
~~~

After:
~~~
[akaris@linux helm]$ bin/helm pull  --repo http://localhost:8080 test --username test --password test
[akaris@linux helm]$ rm -f test-0.1.0.tgz 
[akaris@linux helm]$ bin/helm pull  --repo http://localhost:8080 test --username test --password test2
Error: looks like "http://localhost:8080" is not a valid chart repository or cannot be reached: failed to fetch http://localhost:8080/index.yaml : 401 Unauthorized
[akaris@linux helm]$ bin/helm pull  --repo http://localhost:8080 test --username test
Error: looks like "http://localhost:8080" is not a valid chart repository or cannot be reached: failed to fetch http://localhost:8080/index.yaml : 401 Unauthorized
[akaris@linux helm]$ bin/helm pull  --repo http://localhost:8080 test
[akaris@linux helm]$ rm -f test-0.1.0.tgz 
~~~

felipecrs commented:
This looks cool!

cndoit18 commented:
No offense, as I understand it, you shouldn't read data in the local configuration when using `--repo` (it may cause additional understanding issues).
As in this comment of mine 
https://github.com/helm/helm/issues/9762#issuecomment-852943160

felipecrs commented:
@cndoit18 what would be your solution for the problem stated on https://github.com/helm/helm/issues/9599 so?

I still believe this is the best way to do it, the same way docker and podman reuses your local defined credentials when you try to pull an image from a repository specifying its url.

cndoit18 commented:
> @cndoit18 what would be your solution for the problem stated on #9599 so?
> 
> I still believe this is the best way to do it, the same way docker and podman reuses your local defined credentials when you try to pull an image from a repository specifying its url.
> 
> > (it may cause additional understanding issues)
> 
> That can be easily solved by writing a proper help description to the parameter.

I've replied in your issue

mattfarina commented:
@andreaskaris could you please fix the conflict

andreaskaris commented:
@mattfarina Done

With current master:
~~~
[akaris@linux helm-upstream]$ git log --oneline | head -1
eb994345 Merge pull request #9871 from mattfarina/logging-for-creds
[akaris@linux helm-upstream]$ make
make: Nothing to be done for 'all'.
[akaris@linux helm-upstream]$ bin/helm pull  --repo http://localhost:8080 test
Error: looks like "http://localhost:8080" is not a valid chart repository or cannot be reached: failed to fetch http://localhost:8080/index.yaml : 401 Unauthorized
~~~

With the fix rebased:
~~~
[akaris@linux helm]$  git log --oneline | head -2
8af97d56 fix(helm): helm pull --repo to use repo creds from helm repos file
eb994345 Merge pull request #9871 from mattfarina/logging-for-creds
[akaris@linux helm]$ make
make: Nothing to be done for 'all'.
[akaris@linux helm]$ bin/helm pull  --repo http://localhost:8080 test
[akaris@linux helm]$ 
~~~

mattfarina commented:
I added this PR and the corresponding issue to the next Helm developer call. I think it's worth discussing to come to a quick resolution so we can see when a change would go in.

mattfarina commented:
Moving this to 3.7.0 as it is a feature request.

yxxhero commented on pkg/repo/repo.go:
```diff
@@ -95,6 +95,28 @@ func (r *File) Get(name string) *Entry {
 }
 
+// Get returns an entry with the given URL if it exists, otherwise returns nil
+func (r *File) GetURL(url string) *Entry {
```
GetURL? do you have a better func name?

mattfarina commented:
@andreaskaris We are looking at this for the 3.8.0 release in January. There's a small amount of feedback on the PR. Would you be able to update it?

felipecrs commented:
@mattfarina I created a follow-up PR addressing the comments: https://github.com/helm/helm/pull/10406

scottrigby commented:
Moving to 3.9.0 as we are cutting the January release today. @felipecrs will do the same with your PR, and we can discuss on an upcoming Helm dev call whether your follow-up PR can supersede this one.

felipecrs commented:
@scottrigby thank you. Whatever is needed to get this fixed. :)

isarns commented:
any updates?

felipecrs commented:
I'm happy to refresh my PR if there's any intention on getting it merged.

isarns commented:
### How This PR Could Impact ArgoCD and Kustomize Integration

#### **The Problem**

Currently, Kustomize does not support authentication for private repositories/registries, as confirmed in https://github.com/kubernetes-sigs/kustomize/issues/4401#issuecomment-1213360696 and [in the docs](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/#long-term-support)

This limitation is compounded by the fact that Kustomize does not accept environment variables, making it impossible to securely use private repositories with ArgoCD—secrets would be exposed to Git.

#### **Example of a Non-Working Kustomization File**

The following configuration does not work in a local setting or with ArgoCD:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

helmCharts:
  - repo: https://raw.githubusercontent.com/my-org/helm-private-repository/master/
    version: 0.0.1
    name: package-name
    namespace: test
    releaseName: test
    valuesFile: "values.yaml"
```

This translates to the following command:

```
helm pull --untar --untardir /<Kustomization file path>/charts/package-name-0.0.1 --repo https://raw.githubusercontent.com/my-org/helm-private-repository/master/ package-name --version 0.0.1
```

Resulting in the error:

```
failed to fetch https://raw.githubusercontent.com/my-org/helm-private-repository/master/index.yaml : 404 Not Found
```

Additionally, Kustomize does not currently support referencing repository names instead of URLs https://github.com/kubernetes-sigs/kustomize/issues/4335, but I am working on a PR to address this.

#### **Another Issue: ArgoCD and Helm Repo Credentials**

ArgoCD does not pass Helm repository credentials to Kustomize applications, as detailed in https://github.com/argoproj/argo-cd/issues/10745. However, if Kustomize gains support for private Helm repositories, ArgoCD would likely update its behavior accordingly.

Even if this does not happen immediately, there is a simple workaround https://github.com/argoproj/argo-cd/issues/16623#issuecomment-2176566772.

### **How This PR Can Drive a Significant Change**

This PR has the potential to unlock a long-requested feature in ArgoCD (dating back to at least 2021):

1. **This PR** → Kustomize will support a newer Helm version.
2. **New Kustomize Version** → Will allow working with private Helm repositories locally.
3. **ArgoCD Updates to Use New Kustomize** → This will enable ArgoCD to work with kustomize and potentially private Helm repositories.
4. **Potential ArgoCD Patch** → ArgoCD may then start passing Helm credentials to Kustomize applications.

To validate this, I have already tested a custom Kustomize build sideloaded into ArgoCD, successfully authenticating to a private Helm repository using the workaround mentioned earlier.

### **Related Issues That This PR May Help Solve**

- https://github.com/argoproj/argo-cd/issues/16623
- https://github.com/argoproj/argo-cd/issues/16894
- https://github.com/argoproj/argo-cd/issues/21257
- https://github.com/argoproj/argo-cd/issues/10745
- https://github.com/kubernetes-sigs/kustomize/issues/5407
- https://github.com/kubernetes-sigs/kustomize/issues/4335
- https://github.com/kubernetes-sigs/kustomize/issues/4381

This PR is an important step toward resolving these long-standing issues and improving the overall integration of ArgoCD and Kustomize with private Helm repositories.

andreaskaris commented:
Sorry for having been MIA on this PR. Seems like @felipecrs  addressed some of the points here already but that conflicts must be resolved ... I'm sure that either he or me can have a look at that. From my experience with another PR, it's best to join the weekly (?) meeting where the helm team discuss PRs - at that point, if it's a useful one, Matt is pretty fast getting the right eyes on them

andreaskaris commented:
(let me try to refresh my PR, add felipecrs's changes, and then join the helm meeting) - I'll report back here

isarns commented:
Thanks, @andreaskaris! I really appreciate it.

andreaskaris commented:
FYI, I put this PR onto Thursday's agenda and will join the community call and ask about the maintainers' opinion

isarns commented:
Any updates?

andreaskaris commented:
@gjenkins8 I rebased this to match your changes, can you review, pls?

andreaskaris commented:
I had joined the helm meeting in Feb and someone promised to look at it, but since then, unfortunately, nothing has happened review wise :(

isarns commented:
Hi @gjenkins8 , I hope you're doing well. Could you please review this PR when you have a moment? Thank you so much!

panteparak commented:
@andreaskaris 
Hi, could you have a look at the conflicts, and hopefully it can be merge by the next milestones.

andreaskaris commented:
O.k.,I rebased it yet again ;-)

panteparak commented:
hello, @mattfarina, @yxxhero, @gjenkins8,
I hope you're doing well, It would be good if you could review and merge this PR if you have some available slot.

cfsnate commented:
+1

benoittgt commented on pkg/action/pull.go:
```diff
@@ -73,6 +74,23 @@ func (p *Pull) SetRegistryClient(client *registry.Client) {
+	// password are set for a pull request, check if the repository credentials
+	// are already stored in helm's repository config.
+	if p.RepoURL != "" {
+		if p.Username == "" && p.Password == "" {
```
We can chain them to avoid the double if?
```go
if p.RepoURL != "" && p.Username == "" && p.Password == "" {
```

benoittgt commented on pkg/action/pull.go:
```diff
@@ -73,6 +74,23 @@ func (p *Pull) SetRegistryClient(client *registry.Client) {
+		if p.Username == "" && p.Password == "" {
+			f, err := repo.LoadFile(p.Settings.RepositoryConfig)
+			if err != nil {
+				return out.String(), err
```
Maybe we could provid a better error like:
```go
    return out.String(), fmt.Errorf("failed to load repository config %q: %w", p.Settings.RepositoryConfig, err)
```

benoittgt commented on pkg/action/pull.go:
```diff
@@ -73,6 +74,23 @@ func (p *Pull) SetRegistryClient(client *registry.Client) {
+				return out.String(), err
+			}
+			if repoEntry := f.GetByURL(p.RepoURL); repoEntry != nil {
+				slog.Info("Using credentials from repository config")
```
I'm wondering if it should on debug level instead

andreaskaris commented:
Rebase done, added your suggestions; ran a quick manual verification.

Result:
```
# /home/akaris/development/helm/bin/helm pull --debug --repo http://127.0.0.1/charts/ brigade
level=DEBUG msg="Using credentials from repository config"
level=WARN msg="unable to find exact version; falling back to closest available version" chart=brigade requested="" selected=1.10.0
level=DEBUG msg="setup up default downloader cache"
```

Thanks! (not sure though if, after more than 4 years, this small patch will ever land ;) )

TerryHowe commented:
This needs a `git cherry-pick 895891eba81e95ad4163e69a7cacfd6e54ddb716` to main and force push.

felipecrs commented:
Still valid.
